### ISSUE TITLE
Heavy optimisation of ICs by storing input-output dictionaries 

### ISSUE THREAD

**Issue by Bobingstern** [UnsignedArduino/LogicalSimulator#206]
If the number of inputs to an IC is small enough and the user verifies that it is "deterministic" i.e one input maps to exactly on output (no flip flops or latches) it will store all the possible input output combinations in a dictionary making it MUCH faster than running the circuit every time the IC gets used. When simulating all the inputs it is essential that the current simulation step $x_i$ and the simulation step $x_{i-(N+1)}$ are equal to avoid propagation delay and buffer gate inconsistencies  that may arise where $N$ is the largest delay of the buffer gates in the IC definition.

**Comment by UnsignedArduino** [comment]
On the IC definition blocks we can record a truth table, but how do we know when the input is done propagating? Like say we set several inputs, then how many cycles do we wait for the IC to finish propagating the signal? What about nested ICs?

**Comment by Bobingstern** [comment]
I explained this, we wait until 

>  it is essential that the current simulation step $x_i$ and the simulation step $x_{i-(N+1)}$ are equal to avoid propagation delay and buffer gate inconsistencies that may arise where $N$ is the largest delay of the buffer gates in the IC definition.

**Comment by Bobingstern** [comment]
We can use a another method that I came up with that I like to call "cached chunk update" where we choose a small chunk of components and treat them as an IC and store the input output pairs although somehow we must figure out if it is deterministic or not

**Comment by UnsignedArduino** [comment]
I'm thinking to find out if an IC definition is deterministic, we can search every input and follow the components and connections, and if it loops back, then it isn't deterministic. In the manage ICs dialog, we can have another button for every IC which says "optimize", and when you click it it opens a dialog saying either "This IC definition appears to be deterministic. Do you want to optimize the IC by creating a truth table cache?" or "This IC definition does not appear to be deterministic, and therefore should probably not be cached. Are you sure want to optimize the IC by creating a truth table cache?" 

Also, I finally understand what you mean by the current simulation steps. :) Basically, after finding all the paths, we just take the longest path and count the number of components in it, which will be the number of update cycles needed for the signal to fully update the IC's output. Then every time we change the inputs while building the truth table cache, we just only need to update it that many times before recording the output.

**Comment by Bobingstern** [comment]
Your understanding is still wrong about the simulation step. Assume that the input is 0001 and the output is 0000 initially. Begin a propagation cycle starting on update tick $x_0$ and continue updating until update tick $x_i$ where the output has been changed (maybe 0010). Now, check update tick $x_{i-(N+1)}$ (where $N$ is defined in the above comments) and see if the current on $x_i$ and that are equal, if not continue ticking until they are meaning that the output has not changed in N ticks

**Comment by UnsignedArduino** [comment]
So I just check all the component's power states to see if they have changed after N,and if it hasn't changed then we are done?

Also, after we build a truth table, do you think we need to check it again and if there is a difference then we show an error?

**Comment by Bobingstern** [comment]
Correct, N should be the largest buffer gate delay PLUS the number of components in the IC. My bad

**Comment by Bobingstern** [comment]
I also propose a simple solution to the non determinism of latch based circuit. Simply store input-current_output-output triples. That way, for something like a SR latch, it will know if the state is on or off since it knows what the current output is. This will increase memory usage but decrease complexity

**Comment by UnsignedArduino** [comment]
We can try to finish https://github.com/UnsignedArduino/LogicalSimulator/issues/231 first, then we can finish this.

**Comment by Bobingstern** [comment]
There is a possibility of converting circuits into a boolean expression and getting a truth table out DIRECTLY using boolean algebra rather than raw computation

**Comment by UnsignedArduino** [comment]
Doesn't seem that hard to convert a bunch of component to a bunch of expressions. We can even do this in a web worker.

**Comment by Bobingstern** [comment]
`math.evaluate("!true == (false && true)")` works just fine, and math.js is fast!
